Feature/calendar/delegation

rullzer opened the pull request:
fixes #7085

rullzer committed:
Add table to handle calendar delegations

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Add DB mapper and Entity

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Use the proxymapper to obtain valid proxy data

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

georgehrke commented on apps/dav/lib/CalDAV/Proxy/Proxy.php:
```diff
@@ -0,0 +1,51 @@
+	protected $permissions;
+
+	public function __construct() {
+		$this->addType('ownerId', 'string');
```
default type is string, so adding a type for ownerId and proxyId is not necessary

rullzer replied:
Yes I know. But sill being explicit is nice IMO

georgehrke commented on apps/dav/lib/Migration/Version1011Date20190806104428.php:
```diff
@@ -0,0 +1,53 @@
+		]);
+
+		$table->setPrimaryKey(['id']);
+		$table->addUniqueIndex(['owner_id', 'proxy_id', 'permissions'], 'dav_cal_proxy_uidx');
```
I guess we should add indices for proxy_id and owner_id, because we use that for querying.

rullzer replied:
Fair enough

georgehrke reviewed (changes requested):
First quick dirty review, will check in depth later

rullzer committed:
Early first stage implementation of the groupset

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

tcitworld commented:
Could you explain what's the motive behind this?

georgehrke commented:
> Could you explain what's the motive behind this?

This will allow you to delegate all your calendars to someone else. Most common use-case is someone sharing their calendars with their assistant/secretary to allow them to send out and respond to invitations on their behalf.

tcitworld commented:
Thanks, wasn't sure if it was the issue that you linked on top.

georgehrke committed:
use principaluri instead of userid, allowing to add delegates for rooms and things

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

!fixup add owner_id and proxy_id as db index, since we use it for querying

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

!fixup don't add ACL for each individual proxy, just use calendar-proxy groups

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

!fixup allow delegation of resources / rooms

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

!fixup fix addIndex call in migration

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

!fixup fix remaining constructor calls of Principal

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

!fixup minor fixes and unit tests

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

georgehrke commented on apps/dav/lib/CalDAV/Calendar.php:
```diff
@@ -174,9 +181,37 @@ public function getACL() {
+		list($prefix, $name) = \Sabre\Uri\split($this->getOwner());
+		$proxies = $this->proxyMapper->getProxiesOf($name);
+
+		foreach ($proxies as $proxy) {
```
Note for myself: Remove, just add principal /calendar-user-write

georgehrke commented:
![61A4B590-7041-4254-9A12-47DCD40AC5CC-2](https://user-images.githubusercontent.com/1250540/63020768-48162880-be9f-11e9-9c09-7d6139cdafa8.jpg)

🎉🎉🎉

georgehrke commented:
tldr of what's left:
- [ ] add ProxyMapper to Resource and Room principal backends
- [ ] tests

georgehrke commented:
Let's see if Drone is happy now.

rullzer commented:
Lets merge this!

alex-s-team commented:
pretty excited about this feature, has this been released already (nextcloud 18 is not showing any delegates in ical)? how can this be activated?

tcitworld commented:
@alex-s-team It's available in Nextcloud 17. It will work with CalDAV clients that support the feature, such as Apple Calendar or BusyCal.

alex-s-team commented:
thanks very much for your answer, we are using apple calendar and there are no delegates showing up. do i have to activate the feature somehow via config or is the principal url different?

tcitworld commented:
As this seems to be a setup issue I would like to ask you to raise your question in the forums (so that we don't spam people in this issue): https://help.nextcloud.com

If you wish support with setup issues from Nextcloud GmbH they offer this as part of the Nextcloud subscription. Learn more about this at https://nextcloud.com/enterprise/

rotdrop commented:
Just for the record: here we find that this does not work with the OSX calendar app: #17053. At least it is not possible to find anyone to delegate the calendar to ...

z3ky commented:
Maybe, the caldav trashbin #26083 #1662 has broken delegation #42051
